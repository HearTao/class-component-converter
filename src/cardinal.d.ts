declare module 'cardinal' {
    interface HighlightOptions {
        jsx?: boolean;
    }

    interface Highlight {
        (code: string, options?: HighlightOptions): string;
    }
    export var highlight: Highlight;
}
