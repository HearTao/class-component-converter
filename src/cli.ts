import * as yargs from 'yargs';
import * as prettier from 'prettier';
import * as fs from 'fs';
import * as path from 'path';
import { convert } from './';

// TODO: color support
// import { highlight } from 'cardinal';

interface Options {
    output?: string;
    // color: boolean;
}

export default async function main(): Promise<void> {
    const args: yargs.Arguments<Options> = yargs
        .strict()
        .usage(`$0 <input> [options]`)
        .option(`output`, {
            alias: `o`,
            type: `string`,
            description: `Output to file`,
            default: undefined
        })
        // .option(`color`, {
        //     type: `boolean`,
        //     description: `Colorful output`,
        //     default: true
        // })
        .version()
        .alias(`v`, `version`)
        .showHelpOnFail(true, `Specify --help for available options`)
        .help(`h`)
        .alias(`h`, `help`).argv;

    const { _, output, color } = args;
    if (0 === _.length) throw makeNoInputProvideError();
    const input: string = _[0];
    const content: string = fs.readFileSync(path.resolve(input), `utf-8`);
    const result: string = convert(content);

    if (undefined === output) {
        const formatted = prettier.format(result, { parser: `typescript` });

        // console.log(color ? highlight(formatted, { jsx: true }) : formatted);
        console.log(formatted);
    } else {
        fs.writeFileSync(output, result, `utf-8`);
    }
}

function makeNoInputProvideError(): Error {
    return new Error(`No input file provide`);
}

main();
