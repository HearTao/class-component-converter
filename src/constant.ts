export const contextProperty = [
    '$slots',
    '$scopedSlots',
    '$refs',
    '$emit',
    '$route',
    '$router',
    '$store'
];

export enum Libs {
    vue = 'vue',
    vueTsx = 'vue-tsx-support',
    vuePropertyDecorator = 'vue-property-decorator',
    vueClassComponent = 'vue-class-component'
}

export const lifecycles = [
    'beforeCreate',
    'created',
    'beforeMount',
    'mounted',
    'beforeUpdate',
    'updated',
    'activated',
    'deactivated',
    'beforeDestroy',
    'destroyed',
    'errorCaptured'
];

export enum Decorators {
    Prop = 'Prop',
    Watch = 'Watch',
    Emit = 'Emit',
    Provide = 'Provide',
    Inject = 'Inject'
}

export enum Identifiers {
    steup = 'steup',
    render = 'render',
    context = 'context',
    props = 'props',
    value = 'value',
    computed = 'computed',
    watch = 'watch',
    $emit = '$emit',
    provide = 'provide',
    inject = 'inject'
}
