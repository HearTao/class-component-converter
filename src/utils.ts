export type NotNull<T> = Exclude<T, undefined | null>;

export function some<T>(
    items: ReadonlyArray<T> | null | undefined,
    cb: (v: T) => boolean
): boolean {
    return !!(items && items.some(cb));
}

export function isDef<T>(v: T | undefined): v is T {
    return v !== undefined && v !== null;
}

export function append<T>(to: T[] | undefined, item: T | undefined): T[] {
    if (!isDef(item)) return to || [];
    if (!isDef(to)) return [item];
    to.push(item);
    return to;
}

export function not<P extends any[]>(
    cb: (...args: P) => boolean
): (...args: P) => boolean {
    return (...args: P) => {
        return !cb(...args);
    };
}

export function or<P extends any[]>(
    ...funcs: Array<(...args: P) => boolean>
): (...args: P) => boolean {
    return (...args: P) => {
        return funcs.some(func => func(...args));
    };
}

export function cast<T, U extends T>(value: T, cb: (v: T) => v is U): U {
    if (!cb(value)) {
        throw new Error('invalid cast');
    }
    return value;
}

export function first<T>(items: ReadonlyArray<T>): T {
    if (!items.length) {
        throw new Error('out of range');
    }
    return items[0];
}

export function assertDef<T>(v: T | undefined): T {
    if (!isDef(v)) {
        throw new Error('must be defined');
    }
    return v;
}

export function pickOut<T, U extends T>(
    items: ReadonlyArray<T>,
    cb: (v: T) => v is U
): [U[], T[]] {
    const r1: T[] = [];
    const r2: U[] = [];
    items.forEach(x => {
        if (cb(x)) {
            r2.push(x);
        } else {
            r1.push(x);
        }
    });
    return [r2, r1];
}
