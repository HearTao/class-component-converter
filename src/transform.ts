import * as ts from 'typescript';
import { some, or } from './utils';
import { isVueClass, isComponentDecorator } from './helper';

export function classNeedTransform(
    checker: ts.TypeChecker,
    node: ts.ClassLikeDeclaration
): boolean {
    return (
        some(
            node.heritageClauses,
            x =>
                x.token === ts.SyntaxKind.ExtendsKeyword &&
                some(x.types, type => isVueClass(type, checker))
        ) &&
        some(node.decorators, decorator =>
            isComponentDecorator(decorator, checker)
        )
    );
}
