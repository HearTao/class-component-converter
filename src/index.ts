import * as ts from 'typescript';
import { append, or, pickOut, isDef, push, match, id, mapDef } from './utils';
import createVHost from './host';
import {
    IdentifierName,
    ClassStateDeclaration,
    ClassMethodDeclaration,
    ClassWatchDeclaration,
    ClassEmitDeclaration,
    ClassPropDeclaration,
    ClassProviderDeclaration,
    ClassInjectDeclaration,
    ClassLifeCycleDeclaration,
    ClassComputedDeclaration,
    ComponentInfo
} from './types';
import {
    isClassComputedDeclaration,
    isClassStateDeclaration,
    isClassMethodDeclaration,
    isClassPropDeclaration,
    isClassInjectionDeclaration,
    isClassProviderDeclaration,
    isRenderFunction,
    isClassWatchDeclaration,
    isClassEemitDeclaration,
    isClassLifeCycleDeclaration,
    isValidComputedDeclaration,
    isGetter,
    skipParens
} from './helper';
import { classNeedTransform } from './transform';
import { contextProperty } from './constant';

function collectClassDeclarationInfo(
    node: ts.ClassDeclaration,
    checker: ts.TypeChecker
): ComponentInfo {
    let render: ClassMethodDeclaration | undefined = undefined;
    const computed = new Map<string, ClassComputedDeclaration>();
    const states: ClassStateDeclaration[] = [];
    const props: ClassPropDeclaration[] = [];
    const methods: ClassMethodDeclaration[] = [];
    const emits: ClassEmitDeclaration[] = [];
    const watchers: ClassWatchDeclaration[] = [];
    const lifecycles: ClassLifeCycleDeclaration[] = [];
    const providers: ClassProviderDeclaration[] = [];
    const injections: ClassInjectDeclaration[] = [];
    const ignored: ts.ClassElement[] = [];

    const pushIgnored = push(ignored);
    node.members.forEach(member => {
        switch (member.kind) {
            case ts.SyntaxKind.PropertyDeclaration:
                return match(member)(
                    isClassStateDeclaration.bind(null, checker),
                    push(states)
                )(isClassPropDeclaration.bind(null, checker), push(props))(
                    isClassProviderDeclaration.bind(null, checker),
                    push(providers)
                )(
                    isClassInjectionDeclaration.bind(null, checker),
                    push(injections)
                )(id, pushIgnored);

            case ts.SyntaxKind.GetAccessor:
            case ts.SyntaxKind.SetAccessor:
                return match(member)(
                    isClassComputedDeclaration.bind(null, checker),
                    mem => {
                        const name = mem.name.text;
                        let value = computed.get(name);
                        if (isGetter(mem)) {
                            value = { ...value, getter: mem };
                        } else {
                            value = { ...value, setter: mem };
                        }
                        computed.set(name, value);
                    }
                )(id, pushIgnored);
            case ts.SyntaxKind.MethodDeclaration:
                return match(member)(
                    isRenderFunction.bind(null, checker),
                    mem => (render = mem)
                )(isClassWatchDeclaration.bind(null, checker), push(watchers))(
                    isClassEemitDeclaration.bind(null, checker),
                    push(emits)
                )(
                    isClassLifeCycleDeclaration.bind(null, checker),
                    push(lifecycles)
                )(isClassMethodDeclaration.bind(null, checker), push(methods))(
                    id,
                    pushIgnored
                );
            default:
                return pushIgnored(member);
        }
    });
    return {
        render,
        computed,
        states,
        props,
        methods,
        emits,
        watchers,
        lifecycles,
        providers,
        injections,
        ignored
    };
}

const stateIdentifier = 'value';
const computedIdentifier = 'computed';

export function classTransformer(
    checker: ts.TypeChecker
): ts.TransformerFactory<ts.SourceFile> {
    return context => {
        const visitor: ts.Visitor = node => {
            switch (node.kind) {
                case ts.SyntaxKind.ClassDeclaration:
                    return classDeclarationVisitor(node as ts.ClassDeclaration);
                case ts.SyntaxKind.PropertyAccessExpression:
                    return propertyAccessExpressionVisitor(
                        node as ts.PropertyAccessExpression
                    );
                case ts.SyntaxKind.Identifier:
                    return identifierVisitor(node as ts.Identifier);
                case ts.SyntaxKind.VariableStatement:
                    return variableStatement(node as ts.VariableStatement);
                default:
                    return ts.visitEachChild(node, visitor, context);
            }
        };
        return n => ts.visitNode(n, visitor);

        function classDeclarationVisitor(declaration: ts.ClassDeclaration) {
            if (classNeedTransform(declaration, checker)) {
                return transformClassDeclaration(declaration);
            }
            return ts.visitEachChild(declaration, visitor, context);
        }

        function propertyAccessExpressionVisitor(
            declaration: ts.PropertyAccessExpression
        ) {
            return transformPropertyAccessExpression(
                ts.visitEachChild(declaration, visitor, context),
                checker
            );
        }

        function identifierVisitor(declaration: ts.Identifier) {
            return transformIdentifier(
                ts.visitEachChild(declaration, visitor, context),
                checker
            );
        }

        function variableStatement(declaration: ts.VariableStatement) {
            return transformVariableStatement(
                ts.visitEachChild(declaration, visitor, context)
            );
        }

        function transformClassStates(
            states: ReadonlyArray<ClassStateDeclaration>
        ): ts.VariableStatement[] {
            return states.map(state =>
                ts.createVariableStatement(
                    undefined,
                    ts.createVariableDeclarationList(
                        [
                            ts.createVariableDeclaration(
                                state.name,
                                state.decl.type,
                                ts.createCall(
                                    ts.createIdentifier(stateIdentifier),
                                    undefined,
                                    state.decl.initializer && [
                                        ts.visitEachChild(
                                            state.decl.initializer,
                                            visitor,
                                            context
                                        )
                                    ]
                                )
                            )
                        ],
                        ts.NodeFlags.Const
                    )
                )
            );
        }

        function transformClassProps(
            props: ReadonlyArray<ClassPropDeclaration>
        ): ts.ParameterDeclaration {
            return ts.createParameter(
                undefined,
                undefined,
                undefined,
                ts.createIdentifier('props'),
                undefined,
                ts.createTypeLiteralNode(
                    props.map(prop =>
                        ts.createPropertySignature(
                            undefined,
                            prop.name.text,
                            prop.decl.questionToken,
                            prop.decl.type,
                            undefined
                        )
                    )
                ),
                undefined
            );
        }

        function transformInstanceMethods() {
            return ts.createParameter(
                undefined,
                undefined,
                undefined,
                ts.createIdentifier('context'),
                undefined,
                undefined,
                undefined
            );
        }

        function transformClassComputedDeclaration(
            computed: Map<string, ClassComputedDeclaration>
        ): ts.VariableStatement[] {
            return Array.from(computed.values())
                .filter(isValidComputedDeclaration)
                .map(comp =>
                    ts.createVariableStatement(
                        undefined,
                        ts.createVariableDeclarationList(
                            [
                                ts.createVariableDeclaration(
                                    comp.getter.name.text,
                                    comp.getter.decl.type,
                                    ts.createCall(
                                        ts.createIdentifier(computedIdentifier),
                                        undefined,
                                        append(
                                            [
                                                ts.createArrowFunction(
                                                    undefined,
                                                    undefined,
                                                    [],
                                                    undefined,
                                                    ts.createToken(
                                                        ts.SyntaxKind
                                                            .EqualsGreaterThanToken
                                                    ),
                                                    ts.visitEachChild(
                                                        comp.getter.decl.body,
                                                        visitor,
                                                        context
                                                    )
                                                )
                                            ],
                                            comp.setter &&
                                                ts.createArrowFunction(
                                                    undefined,
                                                    undefined,
                                                    comp.setter.decl.parameters,
                                                    undefined,
                                                    ts.createToken(
                                                        ts.SyntaxKind
                                                            .EqualsGreaterThanToken
                                                    ),
                                                    ts.visitEachChild(
                                                        comp.setter.decl.body,
                                                        visitor,
                                                        context
                                                    )
                                                )
                                        )
                                    )
                                )
                            ],
                            ts.NodeFlags.Const
                        )
                    )
                );
        }

        function transformClassLifeCycleName(name: string): string {
            return `on${name[0].toUpperCase()}${name.slice(1)}`;
        }

        function transformClassLifeCycleDeclaration(
            lifecycles: ReadonlyArray<ClassLifeCycleDeclaration>
        ): ts.ExpressionStatement[] {
            return lifecycles.map(lifecycle =>
                ts.createExpressionStatement(
                    ts.createCall(
                        ts.createIdentifier(
                            transformClassLifeCycleName(lifecycle.name.text)
                        ),
                        undefined,
                        [
                            ts.createArrowFunction(
                                undefined,
                                undefined,
                                [],
                                undefined,
                                ts.createToken(
                                    ts.SyntaxKind.EqualsGreaterThanToken
                                ),
                                ts.visitEachChild(
                                    lifecycle.decl.body,
                                    visitor,
                                    context
                                )
                            )
                        ]
                    )
                )
            );
        }

        function transformClassMethodDeclaration(
            methods: ReadonlyArray<ClassMethodDeclaration>
        ): ts.VariableStatement[] {
            return methods.map(method =>
                ts.createVariableStatement(
                    undefined,
                    ts.createVariableDeclarationList(
                        [
                            ts.createVariableDeclaration(
                                method.name,
                                undefined,
                                ts.createArrowFunction(
                                    undefined,
                                    method.decl.typeParameters,
                                    method.decl.parameters,
                                    undefined,
                                    ts.createToken(
                                        ts.SyntaxKind.EqualsGreaterThanToken
                                    ),
                                    ts.visitEachChild(
                                        method.decl.body,
                                        visitor,
                                        context
                                    )
                                )
                            )
                        ],
                        ts.NodeFlags.Const
                    )
                )
            );
        }

        function transformClassWatchDeclaration(
            watchers: ReadonlyArray<ClassWatchDeclaration>
        ): ts.ExpressionStatement[] {
            return watchers.map(watcher => {
                return ts.createExpressionStatement(
                    ts.createCall(ts.createIdentifier('watch'), undefined, [
                        watcher.watch,
                        ts.createArrowFunction(
                            undefined,
                            undefined,
                            watcher.decl.parameters,
                            undefined,
                            ts.createToken(
                                ts.SyntaxKind.EqualsGreaterThanToken
                            ),
                            ts.visitEachChild(
                                watcher.decl.body,
                                visitor,
                                context
                            )
                        )
                    ])
                );
            });
        }

        function transformClassEmitDeclaration(
            emits: ReadonlyArray<ClassEmitDeclaration>
        ): ts.VariableStatement[] {
            return emits.map(emit => {
                const parameters = emit.decl.parameters;
                const body = ts.visitEachChild(
                    emit.decl.body,
                    visitor,
                    context
                );
                const stmts: ReadonlyArray<ts.Statement> = body
                    ? body.statements
                    : [];
                const [returnArgs, others] = pickOut(
                    stmts,
                    ts.isReturnStatement
                );
                const args = [
                    ts.createStringLiteral(emit.emit),
                    ...returnArgs.map(x => x.expression).filter(isDef),
                    ...parameters.map(x => x.name).filter(ts.isIdentifier)
                ];
                return ts.createVariableStatement(
                    undefined,
                    ts.createVariableDeclarationList(
                        [
                            ts.createVariableDeclaration(
                                emit.name,
                                undefined,
                                ts.createArrowFunction(
                                    undefined,
                                    undefined,
                                    parameters,
                                    undefined,
                                    ts.createToken(
                                        ts.SyntaxKind.EqualsGreaterThanToken
                                    ),
                                    ts.createBlock([
                                        ...others,
                                        ts.createExpressionStatement(
                                            ts.createCall(
                                                ts.createPropertyAccess(
                                                    ts.createIdentifier(
                                                        'context'
                                                    ),
                                                    ts.createIdentifier('$emit')
                                                ),
                                                undefined,
                                                args
                                            )
                                        )
                                    ])
                                )
                            )
                        ],
                        ts.NodeFlags.Const
                    )
                );
            });
        }

        function transformClassProviderDeclaration(
            providers: ReadonlyArray<ClassProviderDeclaration>
        ): ts.ExpressionStatement[] {
            return [
                ts.createExpressionStatement(
                    ts.createCall(ts.createIdentifier('provide'), undefined, [
                        ts.createObjectLiteral(
                            providers.map(provider =>
                                ts.createPropertyAssignment(
                                    provider.provide,
                                    ts.visitEachChild(
                                        provider.decl.initializer,
                                        visitor,
                                        context
                                    )
                                )
                            )
                        )
                    ])
                )
            ];
        }

        function transformClassInjectionDeclaration(
            injections: ReadonlyArray<ClassInjectDeclaration>
        ): ts.VariableStatement[] {
            return injections.map(injection =>
                ts.createVariableStatement(
                    undefined,
                    ts.createVariableDeclarationList(
                        [
                            ts.createVariableDeclaration(
                                injection.name,
                                injection.decl.type,
                                ts.createCall(
                                    ts.createIdentifier('inject'),
                                    undefined,
                                    [ts.createStringLiteral(injection.inject)]
                                )
                            )
                        ],
                        ts.NodeFlags.Const
                    )
                )
            );
        }

        function transformClassDeclarationReturn<T extends IdentifierName>(
            nodes: T[]
        ): [ts.ReturnStatement] {
            return [
                ts.createReturn(
                    ts.createObjectLiteral(
                        nodes.map(node =>
                            ts.createShorthandPropertyAssignment(node.name)
                        )
                    )
                )
            ];
        }

        function transformRenderDeclaration(render: ClassMethodDeclaration) {
            // TODO: transform tsx
            return render.decl;
        }

        function transformClassDeclaration(
            node: ts.ClassDeclaration
        ): ts.VariableStatement {
            const {
                render,
                computed,
                states,
                props,
                methods,
                emits,
                watchers,
                lifecycles,
                providers,
                injections,
                ignored
            } = collectClassDeclarationInfo(node, checker);

            return ts.createVariableStatement(
                undefined,
                ts.createVariableDeclarationList(
                    [
                        ts.createVariableDeclaration(
                            node.name!,
                            undefined,
                            ts.createObjectLiteral(
                                append(
                                    [
                                        ts.createMethod(
                                            undefined,
                                            undefined,
                                            undefined,
                                            'steup',
                                            undefined,
                                            undefined,
                                            [
                                                transformClassProps(props),
                                                transformInstanceMethods()
                                            ],
                                            undefined,
                                            ts.createBlock([
                                                ...transformClassInjectionDeclaration(
                                                    injections
                                                ),
                                                ...transformClassStates(states),
                                                ...transformClassEmitDeclaration(
                                                    emits
                                                ),
                                                ...transformClassMethodDeclaration(
                                                    methods
                                                ),
                                                ...transformClassComputedDeclaration(
                                                    computed
                                                ),
                                                ...transformClassLifeCycleDeclaration(
                                                    lifecycles
                                                ),
                                                ...transformClassWatchDeclaration(
                                                    watchers
                                                ),
                                                ...transformClassProviderDeclaration(
                                                    providers
                                                ),
                                                ...transformClassDeclarationReturn(
                                                    [
                                                        ...injections,
                                                        ...states,
                                                        ...methods,
                                                        ...Array.from(
                                                            computed.keys()
                                                        ).map(x => ({
                                                            name: ts.createIdentifier(
                                                                x
                                                            )
                                                        }))
                                                    ]
                                                )
                                            ])
                                        )
                                    ],
                                    render && transformRenderDeclaration(render)
                                )
                            )
                        )
                    ],
                    ts.NodeFlags.Const
                )
            );
        }

        function transformPropertyAccessName(
            node: ts.PropertyAccessExpression,
            declaration: ts.Declaration | undefined
        ) {
            if (contextProperty.includes(node.name.text)) {
                return ts.createPropertyAccess(
                    ts.createIdentifier('context'),
                    node.name
                );
            }
            if (declaration && ts.isClassElement(declaration)) {
                if (
                    or(
                        isClassComputedDeclaration.bind(null, checker),
                        isClassStateDeclaration.bind(null, checker),
                        isClassMethodDeclaration.bind(null, checker),
                        isClassInjectionDeclaration.bind(null, checker)
                    )(declaration)
                ) {
                    return node.name;
                } else if (isClassPropDeclaration(checker, declaration)) {
                    return ts.createPropertyAccess(
                        ts.createIdentifier('props'),
                        node.name
                    );
                }
            }

            return node;
        }

        function trackOriginalFromAssignment(
            node: ts.VariableDeclaration
        ): ts.Declaration | undefined {
            while (node && ts.isVariableDeclaration(node) && node.initializer) {
                const symbol = checker.getSymbolAtLocation(node.initializer);
                if (symbol && symbol.valueDeclaration) {
                    if (ts.isVariableDeclaration(symbol.valueDeclaration)) {
                        node = symbol.valueDeclaration;
                    } else {
                        return node;
                    }
                }
            }
            return undefined;
        }

        function transformPropertyAccessExpression(
            node: ts.PropertyAccessExpression,
            checker: ts.TypeChecker
        ): ts.Node {
            if (node.expression.kind === ts.SyntaxKind.ThisKeyword) {
                return transformPropertyAccessName(
                    node,
                    mapDef(
                        checker.getSymbolAtLocation(node),
                        symbol => symbol.valueDeclaration
                    )
                );
            }

            const symbol = checker.getSymbolAtLocation(node.expression);
            if (
                symbol &&
                symbol.valueDeclaration &&
                ts.isVariableDeclaration(symbol.valueDeclaration)
            ) {
                const original = trackOriginalFromAssignment(
                    symbol.valueDeclaration
                );
                if (
                    original &&
                    ts.isVariableDeclaration(original) &&
                    original.initializer &&
                    original.initializer.kind === ts.SyntaxKind.ThisKeyword
                ) {
                    return transformPropertyAccessName(
                        node,
                        mapDef(
                            checker.getSymbolAtLocation(original.initializer),
                            symbol =>
                                mapDef(
                                    symbol.members &&
                                        symbol.members.get(node.name
                                            .text as ts.__String),
                                    s => s.valueDeclaration
                                )
                        )
                    );
                }
            }

            return node;
        }

        function transformIdentifier(
            node: ts.Identifier,
            checker: ts.TypeChecker
        ): ts.Node {
            const symbol = checker.getSymbolAtLocation(node);
            if (symbol && symbol.valueDeclaration) {
                const declaration = symbol.valueDeclaration;
                if (
                    ts.isClassElement(declaration) &&
                    declaration.name !== node
                ) {
                    return transformIdentifierMembers(node, declaration);
                } else if (
                    ts.isBindingElement(declaration) &&
                    ts.isObjectBindingPattern(declaration.parent) &&
                    ts.isVariableDeclaration(declaration.parent.parent) &&
                    declaration.parent.parent.name === declaration.parent &&
                    declaration.parent.parent.initializer &&
                    skipParens(declaration.parent.parent.initializer).kind ===
                        ts.SyntaxKind.ThisKeyword
                ) {
                    const typeOfObjectLiteral = checker.getTypeAtLocation(
                        declaration.parent.parent.initializer
                    );
                    const originalSymbol =
                        typeOfObjectLiteral &&
                        checker.getPropertyOfType(
                            typeOfObjectLiteral,
                            node.text
                        );
                    if (
                        originalSymbol &&
                        originalSymbol.valueDeclaration &&
                        ts.isClassElement(originalSymbol.valueDeclaration)
                    ) {
                        return transformIdentifierMembers(
                            node,
                            originalSymbol.valueDeclaration
                        );
                    }
                }
            }

            return node;
        }

        function transformVariableStatement(node: ts.VariableStatement) {
            const list = node.declarationList.declarations
                .map(transformVariableDeclaration)
                .filter(isDef);
            if (list.length) {
                return ts.createVariableStatement(
                    node.modifiers,
                    ts.createVariableDeclarationList(
                        list,
                        node.declarationList.flags
                    )
                );
            }
            return ts.createEmptyStatement();
        }

        function transformVariableDeclaration(node: ts.VariableDeclaration) {
            if (
                node.initializer &&
                skipParens(node.initializer).kind === ts.SyntaxKind.ThisKeyword
            ) {
                return undefined;
            }
            return node;
        }

        function transformIdentifierMembers(
            node: ts.Identifier,
            declaration: ts.ClassElement
        ) {
            if (
                or(
                    isClassComputedDeclaration.bind(null, checker),
                    isClassStateDeclaration.bind(null, checker)
                )(declaration)
            ) {
                return ts.createPropertyAccess(
                    ts.createIdentifier(node.text),
                    ts.createIdentifier('value')
                );
            } else if (
                !(
                    ts.isPropertyAccessExpression(node.parent) &&
                    node.parent.name === node
                ) &&
                isClassPropDeclaration(checker, declaration)
            ) {
                return ts.createPropertyAccess(
                    ts.createIdentifier('props'),
                    ts.createIdentifier(node.text)
                );
            }
            return node;
        }
    };
}

export function convert(code: string): string {
    const host = createVHost();

    const filename = 'mod.tsx';
    host.writeFile(filename, code, false);

    const program = ts.createProgram(
        [filename],
        {
            jsx: ts.JsxEmit.Preserve,
            experimentalDecorators: true,
            target: ts.ScriptTarget.Latest
        },
        host
    );

    const checker = program.getTypeChecker();
    const sourceFile = program.getSourceFile(filename);

    const result = ts.transform(sourceFile!, [classTransformer(checker)]);

    const printer = ts.createPrinter();
    const newCode = printer.printFile(result.transformed[0]);
    return newCode;
}
