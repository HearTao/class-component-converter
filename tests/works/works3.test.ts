import { convert } from '../../src';
import * as prettier from 'prettier';

const code = `
import { Vue, Component, Prop }  from 'vue-property-decorator'

@Component
export default class YourComponent extends Vue {
    @Prop(Number) readonly propsA: number | undefined
}
`;

test(`snap`, () => {
    const ret = prettier.format(convert(code), { parser: `typescript` });
    expect(ret).toMatchSnapshot();
});
