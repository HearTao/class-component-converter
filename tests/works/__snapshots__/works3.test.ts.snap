// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snap 1`] = `
"import { Vue, Component, Prop } from \\"vue-property-decorator\\";
const YourComponent = {
  steup(
    props: {
      propsA: number | undefined;
    },
    context
  ) {
    return {};
  }
};
"
`;
